src/hooks/useAuth.ts

```diff
@@ -1,35 +1,123 @@
-import { useEffect, useState } from 'react';
+import { useSelector } from 'react-redux';
+import { useLazyQuery } from '@apollo/client';
+import { useFlags } from 'launchdarkly-react-client-sdk';
 
-const useAuth = (auth: any): [boolean, any, () => Promise<void>] => {
-  const [authenticated, setAuthenticated] = useState(false);
-  const [user, setUser] = useState(null);
+import {
+  BASIC_USER_PROD,
+  GOVTEAM_DEV,
+  GOVTEAM_PROD,
+  TRB_ADMIN_DEV,
+  TRB_ADMIN_PROD
+} from 'constants/jobCodes';
+import GetSystemIntakeGrbReviewersQuery from 'queries/GetSystemIntakeGrbReviewersQuery';
+import {
+  GetSystemIntakeGrbReviewers,
+  GetSystemIntakeGrbReviewersVariables
+} from 'queries/types/GetSystemIntakeGrbReviewers';
+import { AppState } from 'reducers/rootReducer';
 
-  const handleLogout = async () => {
-    const isAuthenticated = await auth.isAuthenticated();
-    if (isAuthenticated) {
-      auth.logout('/');
-    } else {
-      document.location.href = '/';
+type UseAuthReturn = {
+  /**
+   * Object with helper functions to check job codes / roles
+   *
+   * Returns undefined if user is not set
+   */
+  user:
+    | {
+        isTrbAdmin: () => boolean;
+        isGrtReviewer: () => boolean;
+        isBasicUser: () => boolean;
+        isGrbReviewer: (systemIntakeId: string) => boolean;
+      }
+    | undefined;
+  /** Returns true if function with query is called and loading */
+  loading: boolean;
+};
+
+/** Hook that returns functions to check for roles / job codes */
+const useAuth = (): UseAuthReturn => {
+  const auth = useSelector((state: AppState) => state.auth);
+  const { groups, euaId } = auth;
+
+  const flags = useFlags();
+
+  const [getGrbReviewers, { data, called, loading }] = useLazyQuery<
+    GetSystemIntakeGrbReviewers,
+    GetSystemIntakeGrbReviewersVariables
+  >(GetSystemIntakeGrbReviewersQuery);
+
+  const isUserSet = auth.isUserSet && !loading;
+
+  /** Check if user is TRB admin */
+  const isTrbAdmin = (): boolean => {
+    if (flags.downgradeTrbAdmin) {
+      return false;
     }
+
+    if (groups.includes(TRB_ADMIN_DEV) || groups.includes(TRB_ADMIN_PROD)) {
+      return true;
+    }
+
+    return false;
   };
 
-  useEffect(() => {
-    auth.isAuthenticated().then((isAuthenticated: any) => {
-      if (isAuthenticated !== authenticated) {
-        setAuthenticated(isAuthenticated);
-      }
-    });
-  });
+  /** Check if user is GrtReviewer */
+  const isGrtReviewer = (): boolean => {
+    if (flags.downgradeGovTeam) {
+      return false;
+    }
 
-  useEffect(() => {
-    if (authenticated) {
-      auth.getUser().then(setUser);
-    } else {
-      setUser(null);
+    if (groups.includes(GOVTEAM_DEV) || groups.includes(GOVTEAM_PROD)) {
+      return true;
     }
```

Comment:
Is this code any different from the function in `src/utils.ts`? https://github.com/CMS-Enterprise/easi-app/blob/5799a691fc039bd598ce21719d77b55fbeb0d90a/src/utils/user.ts#L22-L32

Perhaps we could use those functions instead of re-defining?

Reply:
Ah thanks for pointing this out - I meant to bring this up in the PR to get input and completely forgot 😅

`useAuth` is essentially the user functions in hook form. I felt like adding the `isGrbReviewer` function made sense, but I couldn't add it to the existing `user` object since it needs `GetSystemIntakeGrbReviewersQuery`.

So I'm wondering if just replacing the `user` object with `useAuth` would make sense. That way we can have one way to check job codes and it would be simpler with a single import.

[Current usage:](https://github.com/CMS-Enterprise/easi-app/blob/main/src/views/GovernanceReviewTeam/index.tsx#L15-L20)
```tsx
const userGroups = useSelector((state: AppState) => state.auth.groups);
const isUserSet = useSelector((state: AppState) => state.auth.isUserSet);
const flags = useFlags();

if (isUserSet) {
  if (user.isGrtReviewer(userGroups, flags)) {
    // conditional code here
  }
}
```

With `useAuth`:
```tsx
const { user } = useAuth();

// Checking if user exists replaces previous `isUserSet`
if (user) {
  if (user.isGrtReviewer()) {
    // conditional code here
  }
}
```

If we do go this route, I would probably want to make the additional changes and get rid of the duplicate code in a second PR so that it's easier to track the changes.

If we don't want to completely replace the `user` object, I can just use the existing `user` functions in `useAuth` instead of redefining them. Or I can update/rename `useAuth` to only contain the `isGrbReviewer` logic to keep it separate, and then use the existing `user` object for the other job codes when needed.

@adamodd thoughts on this?

Reply:
For now just reference the existing `isGrtReviewer`.

For the scope of this PR I don't have firm thoughts on the similarities between `user`/`useAuth`. I wouldn't expect utils like those to be figured out in one shot with the code paths effected. I have noticed the details pointed out, but I don't want to get caught up with that yet. I'm waiting to see how this new code gets applied more, before I have stronger opinions on recomposing or streamlining the code pattern.

I think for now just tidying up redundant code is good enough.

Reply:
I know I proposed it as an option but after revisiting the code, it feels confusing to keep both `useAuth` and the original `user` object since they have the same purpose and functionality (other than the added `isGrbReviewer` function). I removed the hook and put the `isGrbReviewer` code in the `GovernanceReviewTeam` component for now.

Reply:
Works for me. I didn't know if you had other working plans for the new `useAuth` to eventually absorb all of `user` and didn't want those types of nuances to hold this ticket back.

Reply:
`useAuth` was already a full replacement for the `user` object - functionality was the same, it just had an additional `isGrbReviewer` function. All that was left was to remove `user` and replace it in the codebase, which I was planning to do in a followup PR so that this wasn't held up.